### PR TITLE
feat: support Keccak with sha3

### PR DESCRIPTION
**Pull request by xJonathanLEI**
Adds a new feature `sha3-keccak` to `alloy-primitives` that when enabled, uses `sha3` for Keccak. This can be desirable as in some cases `sha3` is faster.

This feature is added in a completely backward-compatible way: `tiny-keccak` is still the default, and downstream libs/apps will _not_ suddenly start using `sha3` by just doing `cargo update`.

### CODE REVIEW

**Comment by xJonathanLEI** [comment]
Yo any update on this?

**Review by onbjerg — APPROVED** [review]
lgtm impl wise, pending either @DaniPopes or @mattsse if we want this in

**Review by mattsse — COMMENTED** [review]
tbh not super familiar with the diff, but do they even produce the same hashes?

**Comment by onbjerg** [comment]
yeah they do produce the same hashes, both crates are impls of https://csrc.nist.gov/pubs/fips/202/final

File: crates/primitives/src/utils/mod.rs
Comment (mattsse) [inline crates/primitives/src/utils/mod.rs:20]: what happens if both asm-keccak and sha3.keccak are enabled? does it compile?

**Review by mattsse — COMMENTED** [review]
not opposed to it, but defer to @DaniPopes 

1 question about activated both feature gated hashers

File: crates/primitives/src/utils/mod.rs
Comment (xJonathanLEI) [reply crates/primitives/src/utils/mod.rs:20]: Yep this is checked on CI I believe:

https://github.com/alloy-rs/core/blob/022c63db87b6bca348dd8e55f86c94a614b99cb6/.github/workflows/ci.yml#L100-L111

**Review by DaniPopes — APPROVED** [review]
Thanks! I've simplified the cfgs by consolidating them into one place

**Pull request merged**
